src/common/errors_js.ml

```diff
@@ -88,11 +88,106 @@ let print_reason_color ~first ~one_line ~color (message: message) =
+let default_style text = (C.Normal C.Default, text)
+let source_fragment_style text = (C.Normal C.Default, text)
+let error_fragment_style text = (C.Normal C.Red, text)
+let line_number_style text = (C.Dim C.Default, text)
+
+(* TODO: Not sure how to handle the usecase when flow reads file content from stdin *)
+let load_file f =
+  let ic = open_in f in
+  let n = in_channel_length ic in
+  let s = Bytes.create n in
+  really_input ic s 0 n;
+  close_in ic;
+  (s)
+
+(* TODO: Add proper support for relative path, for example ../../.. *)
+let relative_path filename =
+  let pwd = (Sys.getcwd ()) ^ "/" in
+  Str.global_replace (Str.regexp_string pwd) "" filename
+
+let highlight_error_in_line line c0 c1 =
+  let prefix = String.sub line 0 c0 in
+  let fragment = String.sub line c0 (c1 - c0) in
+  let suffix = String.sub line c1 ((String.length line) - c1) in
+  [
+    source_fragment_style prefix;
+    error_fragment_style fragment;
+    source_fragment_style suffix;
+  ]
+
+let print_file_at_location main_file loc s = Loc.(
+  let l0 = loc.start.line in
+  let l1 = loc._end.line in
+  let c0 = loc.start.column in
+  let c1 = loc._end.column in
+  match loc.source with
+    | Some filename ->
+      let content = load_file filename in
+      let lines = Str.split (Str.regexp "^") content in
+      let code_line = if (List.length lines) >= l0 then List.nth lines (l0 - 1) else "" in
+      let code_line = Str.global_replace (Str.regexp_string "\n") "" code_line in
+      let highlighted_line = if (l1 == l0) && (String.length code_line) > 0
+        then highlight_error_in_line code_line c0 c1
+        else [source_fragment_style code_line]
+      in
+      let padding_size = max 0 (50 - (String.length code_line)) in
+      let padding = String.make padding_size ' ' in
+      let see_another_file = if filename == main_file then [(default_style "")] else
+        [
+          default_style ". See: ";
+          file_location_style (relative_path filename);
+          file_location_style (":" ^ (string_of_int l0))
+        ]
+      in
+      line_number_style (Printf.sprintf "%3d: " l0) ::
+      highlighted_line @
+      [default_style (Printf.sprintf "%s ← %s" padding s)] @
+      see_another_file @
+      [default_style "\n"]
+
+    | None -> [default_style (Printf.sprintf "%s\n" s)]
+)
+
+let print_message_nice main_file message =
+  let loc, s = to_pp message in
+  print_file_at_location main_file loc s
+
+let print_error_header message = Loc.(
+  let loc, _ = to_pp message in
+  let filename = match loc.source with
+    | Some filename -> filename
+    | None -> ""
+  in
+  let relfilename = relative_path filename in
+  [
+    file_location_style (Printf.sprintf "%s" relfilename);
+    default_style "\n"
+  ]
+)
+
+let append_comment blame comment =
+  match blame with
+  | BlameM(loc, s) ->
+    (match comment with
+    | "Error:" -> BlameM(loc, s) (* Almost everywhere we have "Error:" that hurts readability *)
```

Comment:
This is OK for now, but we need to remember on the error producing side to reserve the string "Error:" for its current purpose. Eventually we'll make the errors more structured so that this special case handling can be checked by the compiler.